Std/Tactic/Alias.lean

```diff
@@ -0,0 +1,165 @@
+def AliasInfo.toString : AliasInfo → String
+  | plain n => s!"**Alias** of `{n}`."
+  | forward n => s!"**Alias** of the forward direction of `{n}`."
+  | reverse n => s!"**Alias** of the reverse direction of `{n}`."
+
+
+/-- Environmant extension for registering aliases -/
+initialize aliasExt : SimpleScopedEnvExtension (Name × AliasInfo) (NameMap AliasInfo) ←
+  registerSimpleScopedEnvExtension {
+    addEntry := fun st (n, i) => st.insert n i
+    initial := {}
+  }
+
+/-- Get the alias information for a name -/
+def getAliasInfo [Monad m] [MonadEnv m] (name : Name) : m (Option AliasInfo) := do
+  return aliasExt.getState (← getEnv) |>.find? name
+
+/-- Set the alias info for a new declaration -/
+def setAliasInfo [MonadEnv m] (info : AliasInfo) (declName : Name) : m Unit :=
+  modifyEnv (aliasExt.addEntry · (declName, info))
+
+/--
+  The command `alias name := target` creates a synonym of `target` with the given name.
+
+  The command `alias ⟨fwd, rev⟩ := target` creates synonyms for the forward and reverse directions
+  of an iff theorem. Use `_` if only one direction is required.
+
+  These commands accept all modifiers and attributes that `def` and `theorem` do.
+ -/
+elab (name := alias) mods:declModifiers "alias " alias:ident " := " name:ident : command =>
+  Command.liftTermElabM do
+    let resolved ← resolveGlobalConstNoOverloadWithInfo name
+    let const ← getConstInfo resolved
+    let declMods ← elabModifiers mods
+    let (declName, _) ← mkDeclName (← getCurrNamespace) declMods alias.getId
+    let decl : Declaration := match const with
+      | Lean.ConstantInfo.thmInfo t =>
+        .thmDecl { t with
+          name := declName
+          value := mkConst resolved (t.levelParams.map mkLevelParam)
+        }
+      | Lean.ConstantInfo.defnInfo c
+      | Lean.ConstantInfo.quotInfo c
+      | Lean.ConstantInfo.inductInfo c
+      | Lean.ConstantInfo.axiomInfo c
+      | Lean.ConstantInfo.opaqueInfo c
+      | Lean.ConstantInfo.ctorInfo c
+      | Lean.ConstantInfo.recInfo c =>
+        .defnDecl { c with
+          name := declName
+          value := mkConst resolved (c.levelParams.map mkLevelParam)
+          hints := .regular 0 -- FIXME
+          safety := if declMods.isUnsafe then .unsafe else .safe
+        }
+    checkNotAlreadyDeclared declName
+    if declMods.isNoncomputable then
+      addDecl decl
+    else
+      addAndCompile decl
+    Lean.addDeclarationRanges declName {
+      range := ← getDeclarationRange (← getRef)
+      selectionRange := ← getDeclarationRange alias
+    }
+    addDocString' declName declMods.docString?
+    Term.applyAttributes declName declMods.attrs
+    let info := match ← getAliasInfo const.name with
+      | some i => i
+      | none => AliasInfo.plain const.name
+    setAliasInfo info declName
+    /- alias doesn't trigger the missing docs linter so we add a default -/
+    if (← findDocString? (← getEnv) declName).isNone then
+      addDocString declName info.toString
```

Comment:
```suggestion
      let mut doc := info.toString
      if let some origDoc ← findDocString? (← getEnv) resolved then
        doc := s!"{doc}\n\n---\n\n{origDoc}"
      addDocString declName doc
```
The alias documentation is nice, but as a user I usually want to see the original docs instead.

Reply:
I'm a bit puzzled here: How would a user know this is an alias? I don't understand how they would come to want to see the original docs. Of course, if you happen to know it's an alias you can just use the original if that's what you want, no?

In any case, I'm happy to include the original docs. What should I do if there's an `inherit_doc`? Should I dig to see if it's an alias of an alias, or should I just pile on the docs?

Reply:
When you make an alias to a definition, a user of that alias is generally using it because the new name is "better". It is very much deliberate that there is no linter to warn if you use an alias by its new name - if you want this effect you use `@[deprecated]` instead. Mathlib has plenty of examples of this, for example `lt_trans` is aliased to `LT.lt.trans` so that people can write `h1.trans h2` instead of `lt_trans h1 h2`.

When you hover over `.trans` there, it is useful to report that this is an alias of `lt_trans`, but if `lt_trans` itself was complicated and confusing and had a big docstring explaining what it means, I think a user would want to see that docstring too on `.trans`.

> In any case, I'm happy to include the original docs. What should I do if there's an inherit_doc? Should I dig to see if it's an alias of an alias, or should I just pile on the docs?

Aliases of aliases should already be resolved. That is, `alias B := A; alias C := B` should have the same effect as `alias B := A; alias C := A`, the logic of which is at least implemented in the `aliasExt`. This does have a slightly unfortunate effect if `B` has a custom docstring since `C` will skip it and show A's docstring instead, but docstrings on aliases are rare anyway. Either `inherit_doc` or a manual docstring should cancel all automated docstring behavior and just show the docstring that was requested; this is included in the code snippet I gave since it is inside the `if docs.isNone then` branch.

Reply:
That clears up my confusion. Thanks!

Note that `aliasExt` does implement that logic but it is not used to resolve the alias yet. Should I implement that as well?

Reply:
I'll fix the resolution step to use `aliasExt` and get to the root alias.